File: 113.path-sum-ii.java
/*
 * @lc app=leetcode id=113 lang=java
 *
 * [113] Path Sum II
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {

    List<List<Integer>> res = new LinkedList<>();
    Stack<Integer> stack = new Stack<>();

    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        if (root != null) fillValidPaths(root, sum);
        return res;
    }

    private void fillValidPaths(TreeNode root, int sum) {
        sum -= root.val;
        stack.push(root.val);
        if (root.left == null && root.right == null && sum == 0) res.add(new LinkedList<>(stack));
        if (root.left != null) fillValidPaths(root.left, sum);
        if (root.right != null) fillValidPaths(root.right, sum);
        stack.pop();
    }
}

File: go/113.path-sum-ii.go
/*
 * @lc app=leetcode id=113 lang=golang
 *
 * [113] Path Sum II
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func pathSum(root *TreeNode, sum int) [][]int {
    if root == nil {
		return [][]int{}
	}
	return pathSumStep(root, sum, []int{})
}

func pathSumStep(root *TreeNode, sum int, path []int) [][]int {
	sum -= root.Val
	path = append(path, root.Val)
	if root.Left == nil && root.Right == nil {
		if sum == 0 {
			path2 := make([]int, len(path))
			copy(path2, path)
			return [][]int{path2}
		}
		return [][]int{}
	}
	res := [][]int{}
	if root.Left != nil {
		res = append(res, pathSumStep(root.Left, sum, path)...)
	}
	if root.Right != nil {
		res = append(res, pathSumStep(root.Right, sum, path)...)
	}
	return res
}
// @lc code=end

